### DISCUSSION TITLE
NeoPool commands results error

### DISCUSSION THREAD

**Discussion by ChristianBlenk** [arendst/Tasmota#18991] [Support request] [answered]
Hi there. In sum i have two problems this this project.
Working with esp32-DevKit(0)
Problem 1
All is working fine so far. I cant turn on the filtrationpump by using NPFiltration 1. Result is always {"NPFiltration":"Error"}

If the pump runs and i send NPFiltration 0, pump stops and went to manual. To start the only workaround is to set NPFiltrationMode to Auto (1).

Problem 2
esp32 berry scripts dont work as shown on [https://tasmota.github.io/docs/NeoPool/#esp32-add-gui-controls-for-filtration-light-and-aux-relais](url)

Any advice?

**Comment by barbudor** [comment]
@ChristianBlenk 

# Problem 1:
Which exact Tasmota version are you using ?

If you are using a build from before Jun 19th, this means that you get no response from the NeoPool => check connection and hardware

If you are using a more recent version, then probably @curzon01 has a problem with his latest change

#Problem 2 
Instead of saying that it doesn't work as expected, please elaborate and details exactly what you have done and what is the result

**Comment by ChristianBlenk** [comment]
Hi,

OK sorry for less information. My english is as bad as coding berry script ;).So i keep my explenations as short as possible.

This is my build.

Program Version 12.5.0.2(tasmota32)
Build Date & Time 2023-05-09T20:26:39
Core/SDK Version 2_0_9/4.4.4.20230310

Core data from NeoPool are sent and received via modbus. The communication is working. All data is present. Eg redox potential or running modes and so on.

As discriped is the documentation, NPFiltration and all other commands in the console response with an error. But sending command NPfiltrationmode 2 for example response with error but is working. The mode switches to automode.

Problem 2
I followed the description to add some buttons to the gui via berry script. This dont work at all.

Added the three *.be files via filemanager, loaded them in berry console without error. 
This is what i get after opening the berry console

Bry: Exception> 'key_error' - Speed

Regards 
Chris
________________________________
Von: Barbudor ***@***.***>
Gesendet: Mittwoch, 28. Juni 2023 17:38
An: arendst/Tasmota
Cc: ChristianBlenk; Mention
Betreff: Re: [arendst/Tasmota] NeoPool commands results error (Discussion #18991) 

@ChristianBlenk
Problem 1:
Which exact Tasmota version are you using ?
If you are using a build from before Jun 19th, this means that you get no response from the NeoPool => check connection and hardware
If you are using a more recent version, then probably @curzon01 has a problem with his latest change
#Problem 2
Instead of saying that it doesn't work as expected, please elaborate and details exactly what you have done and what is the result

**Comment by barbudor** [comment]
# Problem 1

> Build Date & Time 2023-05-09T20:26:39

So you are using a build before the last change
Which means that the command `NPFiltration` return an error in case of communication problem
If you confirm that
- the command is actually taken into account by the device => you don't have problem talking to the device
- the other parameters are returned properly from the device => you don't have problem receiving from the device

Then it is may be that your device is slow (can't answer properly right after receiving a command)

There was a recent attempt to fix that on June 19th so I suggest that you upgrade either to release 13.0 or to latest 13.0.x development version and report if it changes anything.

# Problem 2

The `Speed` field seems to be used in script `neopoolgui.be` in line
```
var speed = tasmota.cmd('NPFiltration')['Speed']
```
So if the command `NPFiltration` with no argument is returning `Error` it is normal that this line fails becaue Berry cannot find the `Speed`  info in the JSON

I suggest that you modify those 2 lines:
```
    var speed = tasmota.cmd('NPFiltration')['Speed']
    var mode = tasmota.cmd('NPFiltrationmode')['NPFiltrationmode']
```
with
```
    var speed = tasmota.cmd('NPFiltration').find('Speed', 'invalid')
    var mode = tasmota.cmd('NPFiltrationmode').find('NPFiltrationmode', 'invalid')
```
which will avoid a crash in the script and set the variable to 'invalid' if not found

But definitively, problem 2 occurs because of problem 1

**Comment by ChristianBlenk** [comment]
Hi,

thanks in advance.

Ill try a new build and report the result

-Chris
________________________________
Von: Barbudor ***@***.***>
Gesendet: Mittwoch, 28. Juni 2023 19:10
An: arendst/Tasmota
Cc: ChristianBlenk; Mention
Betreff: Re: [arendst/Tasmota] NeoPool commands results error (Discussion #18991) 

Problem 1
>
> Build Date & Time 2023-05-09T20:26:39

So you are using a build before the last change
Which means that the command NPFiltration return an error in case of communication problem
If you confirm that
the command is actually taken into account by the device => you don't have problem talking to the device
the other parameters are returned properly from the device => you don't have problem receiving from the device
Then it is may be that your device is slow (can't answer properly right after receiving a command)
There was a recent attempt to fix that on June 19th so I suggest that you upgrade either to release 13.0 or to latest 13.0.x development version and report if it changes anything.
Problem 2
The Speed field seems to be used in script neopoolgui.be in line

var speed = tasmota.cmd('NPFiltration')['Speed']

So if the command NPFiltration with no argument is returning Error it is normal that this line fails becaue Berry cannot find the Speed info in the JSON
I suggest that you modify those 2 lines:

    var speed = tasmota.cmd('NPFiltration')['Speed']
    var mode = tasmota.cmd('NPFiltrationmode')['NPFiltrationmode']

with

    var speed = tasmota.cmd('NPFiltration').find('Speed', 'invalid')
    var mode = tasmota.cmd('NPFiltrationmode').find('NPFiltrationmode', 'invalid')

which will avoid a crash in the script and set the variable to 'invalid' if not found
But definitively, problem 2 occurs because of problem 1

**Reply by ChristianBlenk** [reply]
@barbudor 
changed the script. Works so far. GUI show the buttons and boxes now. But only the change Mode is working.

**Comment by curzon01** [comment]
@barbudor excellent analysis - I should catch a few more exceptions within the scripts ;-)

@ChristianBlenk important info missing is which device you are using with the Neopool driver?

There may still be a communication problem, e.g. the redox value is only the read of a single register, switching the filtration requires several successive read/write accesses.

You can get more information about this with Weblog 4, if that is not enough, you can translate Tasmota with activated sensor logging (#define DEBUG_TASMOTA_SENSOR).

**Reply by ChristianBlenk** [reply]
@curzon01 
What device do you mean? Oxilife or Tasmota Host?
salting - Oxilife
Tasmota Host - Hailege ESP32 ESP-32D ESP-WROOM-32 Development
Modbus - DSD TECH SH-U12 RS485 to TTL 5V Brett with MAX13487

**Comment by curzon01** [comment]
Independent of the NPFiltration error result:

NPFiltration with switch parameter always switchs to manual mode. You cant switch filtration in Auto mode, not even on the unit.

**Reply by ChristianBlenk** [reply]
@curzon01 
Yes i know that NPFiltration 0 ends in manual off. But if i send command NPFiltration 1 via console it ends in error and nothing happend.

**Comment by curzon01** [comment]
To be exact: NPFiltration 0|1 ends in manual mode, it doesn't matter if you turn it off or on.

I have already understood that NPFiltration 1 returns "error". 

Commonly "Error" is returning due to a communication problems. This can be hardware based (interface) or your device is not responding to the Modbus register, but Oxilife should work.

Unlike all other commands like NPFiltrationMode, NPRedox, which all read and write only one simple Modbus register, the command NPFiltration needs several read/write operations on different registers. 

Unfortunately, it is not possible to deduce from the functionality of the simple commands that the communication works at all. If there is a communication problem, the processing of a single register may just work, but the processing of several registers, as required for NPFiltration, will no longer work.

Weblog 4 and possibly compiling Tasmota with active sensor debugging shows whether there are problems here or not.

**Comment by curzon01** [comment]
As long as cmnd return error, communication is broken in one way or another and changes to the script are workarounds around the real problem.

Weblog 4 and/or sensor debugging helps...

**Reply by ChristianBlenk** [reply]
Here is a snip from console after pressing button Filtration with Weblog 4

16:32:57.601 HTP: Main Menu
16:32:57.672 SRC: Berry
16:32:57.674 CMD: Grp 0, Cmd 'NPFILTRATION', Idx 1, Len 0, Pld -99, Data ''
16:32:57.733 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltration":"Error"}
16:32:57.757 SRC: Berry
16:32:57.759 CMD: Grp 0, Cmd 'NPFILTRATIONMODE', Idx 1, Len 0, Pld -99, Data ''
16:32:57.828 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltrationmode":"Error"}
16:32:57.856 BRY: GC from 24916 to 15799 bytes, objects freed 33/218 (in 1 ms) - slots from 198/228 to 164/228
16:32:57.858 BRY: GC from 26468 to 16739 bytes, objects freed 11/218 (in 0 ms) - slots from 164/228 to 164/228
16:32:57.860 BRY: GC from 27592 to 17483 bytes, objects freed 8/218 (in 0 ms) - slots from 164/228 to 164/228
16:32:57.872 BRY: GC from 20817 to 14345 bytes, objects freed 4/215 (in 1 ms) - slots from 164/228 to 163/213
16:33:06.035 SRC: Berry
16:33:06.037 CMD: Grp 0, Cmd 'NPFILTRATION', Idx 1, Len 1, Pld 2, Data '2'
16:33:06.095 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltration":"Error"}

16:36:25.940 CMD: NPFiltrationmode 0
16:36:25.942 SRC: WebConsole from 192.168.122.179
16:36:25.944 CMD: Grp 0, Cmd 'NPFILTRATIONMODE', Idx 1, Len 1, Pld 0, Data '0'
16:36:26.027 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltrationmode":"Error"}
16:36:32.241 HTP: Consoles
16:36:34.225 HTP: Main Menu
16:36:34.297 SRC: Berry
16:36:34.299 CMD: Grp 0, Cmd 'NPFILTRATION', Idx 1, Len 0, Pld -99, Data ''
16:36:34.358 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltration":"Error"}
16:36:34.384 SRC: Berry
16:36:34.386 CMD: Grp 0, Cmd 'NPFILTRATIONMODE', Idx 1, Len 0, Pld -99, Data ''
16:36:34.446 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltrationmode":"Error"}
16:36:34.474 BRY: GC from 24821 to 15008 bytes, objects freed 41/218 (in 1 ms) - slots from 206/228 to 164/228
16:36:34.479 BRY: GC from 26172 to 16281 bytes, objects freed 17/218 (in 0 ms) - slots from 169/228 to 164/228
16:36:34.482 BRY: GC from 26532 to 16947 bytes, objects freed 9/218 (in 0 ms) - slots from 164/228 to 164/228
16:36:34.484 BRY: GC from 27398 to 17750 bytes, objects freed 7/218 (in 0 ms) - slots from 164/228 to 164/228
16:36:34.507 BRY: GC from 17750 to 14345 bytes, objects freed 2/215 (in 0 ms) - slots from 164/228 to 163/228
16:36:35.054 WIF: Checking connection...

16:37:12.912 CMD: NPFiltartionmode 1
16:37:12.915 SRC: WebConsole from 192.168.122.179
16:37:12.917 CMD: Grp 0, Cmd 'NPFILTARTIONMODE', Idx 1, Len 1, Pld 1, Data '1'
16:37:12.930 MQT: stat/PoolHydro_52A200/RESULT = {"Command":"Unknown"}
16:37:15.321 WIF: Checking connection...
16:37:29.447 SRC: Berry
16:37:29.449 CMD: Grp 0, Cmd 'NPFILTRATIONMODE', Idx 1, Len 1, Pld 1, Data '1'
16:37:29.542 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltrationmode":"Error"}
16:37:35.293 WIF: Checking connection...
16:37:48.311 BRY: GC from 24941 to 14940 bytes, objects freed 28/213 (in 1 ms) - slots from 188/228 to 161/228
16:37:55.534 WIF: Checking connection...

**Comment by curzon01** [comment]
The line `{"Command":"Unknown"}` after `NPFiltartionmode 1` shows, that the driver briefly does not retrived date from your Oxilife, which means there is for a short time no system knowning connected - all this points to communication problems.

But does not contain more info.
You have to do it with 
`#define DEBUG_TASMOTA_SENSOR`
that logs the Modbus traffic end errors

**Reply by ChristianBlenk** [reply]
Okay. Going to 13.0.0 with
`#define DEBUG_TASMOTA_SENSOR`

**Comment by ChristianBlenk** [comment]
Okay here is the log. Updated to 13.0.0 with debug_tasmota_Sensor

Salting Device Oxilife 1 (Hobby Pool, Sugar Valey)
Hydro Mod. V01.0A
Screen V4.0
Energy V3.57 May23 2020 01:08:10

20:32:21.503 NEO: addr 0x0002 read data error 159
20:32:22.761 NEO: modbus send(1, 4, 0x0022, 2)
20:32:22.964 NEO: modbus send(1, 4, 0x006A, 9)
20:32:23.269 NEO: modbus send(1, 4, 0x0100, 17)
20:32:23.488 NEO: modbus send(1, 4, 0x0206, 20)
20:32:23.659 SRC: Berry
20:32:23.661 CMD: Grp 0, Cmd 'NPFILTRATIONMODE', Idx 1, Len 1, Pld 0, Data '0'
20:32:23.668 NEO: NeoPoolWriteRegister(0x0411, 1) = [0x0000]
20:32:23.706 NEO: addr 0x0411 read data error 9
20:32:23.723 NEO: addr 0x0411 read data error 9
20:32:23.740 NEO: addr 0x0411 read data error 9
20:32:23.749 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltrationmode":"Error"}
20:32:24.968 NEO: modbus send(1, 4, 0x0206, 20)
20:32:25.220 NEO: modbus send(1, 4, 0x0300, 7)
20:32:25.473 NEO: modbus send(1, 4, 0x0408, 14)
20:32:25.723 NEO: modbus send(1, 4, 0x0500, 16)
20:32:25.978 NEO: modbus send(1, 4, 0x0600, 8)
20:32:26.296 NEO: addr 0x0002 read data error 159
20:32:26.313 NEO: addr 0x0002 read data error 159
20:32:26.330 NEO: addr 0x0002 read data error 159
20:32:27.495 NEO: modbus send(1, 4, 0x0600, 8)
20:32:27.747 NEO: modbus send(1, 4, 0x0022, 2)
20:32:28.005 NEO: modbus send(1, 4, 0x006A, 9)
20:32:28.255 NEO: modbus send(1, 4, 0x0100, 17)
20:32:28.507 NEO: modbus send(1, 4, 0x0206, 20)
20:32:28.815 NEO: modbus send(1, 4, 0x0300, 7)
20:32:28.982 NEO: modbus send(1, 4, 0x0408, 14)
20:32:29.234 NEO: modbus send(1, 4, 0x0500, 16)
20:32:29.484 NEO: modbus send(1, 4, 0x0600, 8)
20:32:29.735 NEO: modbus send(1, 4, 0x0022, 2)
20:32:29.988 NEO: modbus send(1, 4, 0x006A, 9)
20:32:30.110 SRC: Berry
20:32:30.112 CMD: Grp 0, Cmd 'NPFILTRATION', Idx 1, Len 1, Pld 2, Data '2'
20:32:30.134 NEO: addr 0x050F read data error 9
20:32:30.151 NEO: addr 0x050F read data error 9
20:32:30.168 NEO: addr 0x050F read data error 9
20:32:30.177 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltration":"Error"}
20:32:31.116 NEO: addr 0x0002 read data error 159
20:32:31.133 NEO: addr 0x0002 read data error 159
20:32:31.150 NEO: addr 0x0002 read data error 159
20:32:31.497 NEO: modbus send(1, 4, 0x006A, 9)
20:32:31.748 NEO: modbus send(1, 4, 0x0100, 17)
20:32:32.005 NEO: modbus send(1, 4, 0x0206, 20)
20:32:32.256 NEO: modbus send(1, 4, 0x0300, 7)
20:32:32.502 WIF: Checking connection...
20:32:32.509 NEO: modbus send(1, 4, 0x0408, 14)
20:32:32.771 NEO: modbus send(1, 4, 0x0500, 16)
20:32:32.972 NEO: modbus send(1, 4, 0x0600, 8)
20:32:33.223 NEO: modbus send(1, 4, 0x0022, 2)
20:32:33.475 NEO: modbus send(1, 4, 0x006A, 9)
20:32:33.726 NEO: modbus send(1, 4, 0x0100, 17)
20:32:33.981 NEO: modbus send(1, 4, 0x0206, 20)
20:32:34.231 NEO: modbus send(1, 4, 0x0300, 7)
20:32:34.485 NEO: modbus send(1, 4, 0x0408, 14)
20:32:34.736 NEO: modbus send(1, 4, 0x0500, 16)
20:32:34.992 NEO: modbus send(1, 4, 0x0600, 8)
20:32:35.248 NEO: modbus send(1, 4, 0x0022, 2)
20:32:35.503 NEO: modbus send(1, 4, 0x006A, 9)
20:32:35.754 NEO: modbus send(1, 4, 0x0100, 17)
20:32:35.818 NEO: addr 0x0002 read data error 159
20:32:35.835 NEO: addr 0x0002 read data error 159
20:32:35.852 NEO: addr 0x0002 read data error 159
20:32:37.181 SRC: Berry
20:32:37.183 CMD: Grp 0, Cmd 'NPFILTRATIONMODE', Idx 1, Len 1, Pld 1, Data '1'
20:32:37.190 NEO: NeoPoolWriteRegister(0x0411, 1) = [0x0001]
20:32:37.227 NEO: addr 0x0411 read data error 9
20:32:37.256 NEO: addr 0x0411 read data error 7
20:32:37.273 NEO: addr 0x0411 read data error 9
20:32:37.282 MQT: stat/PoolHydro_52A200/RESULT = {"NPFiltrationmode":"Error"}
20:32:37.334 NEO: modbus send(1, 4, 0x0100, 17)
20:32:37.504 NEO: modbus send(1, 4, 0x0206, 20)
20:32:37.753 NEO: modbus send(1, 4, 0x0300, 7)
20:32:38.010 NEO: modbus send(1, 4, 0x0408, 14)
20:32:38.262 NEO: modbus send(1, 4, 0x0500, 16)
20:32:38.508 NEO: modbus send(1, 4, 0x0600, 8)
20:32:38.766 NEO: modbus send(1, 4, 0x0022, 2)
20:32:38.968 NEO: modbus send(1, 4, 0x006A, 9)
20:32:39.220 NEO: modbus send(1, 4, 0x0100, 17)
20:32:39.477 NEO: modbus send(1, 4, 0x0206, 20)
20:32:39.727 NEO: modbus send(1, 4, 0x0300, 7)
20:32:39.859 NEO: addr 0x0610 read data error 9
20:32:39.876 NEO: addr 0x0610 read data error 9
20:32:39.893 NEO: addr 0x0610 read data error 9
20:32:39.910 NEO: addr 0x0611 read data error 1
20:32:39.928 NEO: addr 0x0611 read data error 1
20:32:39.945 NEO: addr 0x0611 read data error 1
20:32:39.962 NEO: addr 0x0612 read data error 1
20:32:39.979 NEO: addr 0x0612 read data error 1
20:32:39.996 NEO: addr 0x0612 read data error 1
20:32:40.016 NEO: NeoPoolReadRegister(0x0613, 1) = [0x0000]
20:32:40.033 NEO: addr 0x0614 read data error 1
20:32:40.050 NEO: addr 0x0614 read data error 1
20:32:40.068 NEO: addr 0x0614 read data error 1
20:32:40.087 NEO: NeoPoolReadRegister(0x0615, 1) = [0x0000]
20:32:40.106 NEO: NeoPoolReadRegister(0x0616, 1) = [0x0000]
20:32:40.125 NEO: NeoPoolReadRegister(0x0617, 1) = [0x0000]
20:32:40.142 NEO: addr 0x0618 read data error 1
20:32:40.159 NEO: addr 0x0618 read data error 1
20:32:40.176 NEO: addr 0x0618 read data error 1
20:32:40.196 NEO: NeoPoolReadRegister(0x0619, 1) = [0x0000]
20:32:40.216 NEO: NeoPoolReadRegister(0x061A, 1) = [0x0000]
20:32:40.235 NEO: NeoPoolReadRegister(0x061B, 1) = [0x0000]
20:32:40.254 NEO: NeoPoolReadRegister(0x061C, 1) = [0x0000]
20:32:40.273 NEO: NeoPoolReadRegister(0x061D, 1) = [0x0000]
20:32:40.292 NEO: NeoPoolReadRegister(0x061E, 1) = [0x0000]
20:32:41.092 NEO: addr 0x0002 read data error 159
20:32:41.109 NEO: addr 0x0002 read data error 159
20:32:41.126 NEO: addr 0x0002 read data error 159
20:32:41.472 NEO: modbus send(1, 4, 0x0300, 7)
20:32:41.663 HTP: Consoles
20:32:41.769 NEO: modbus send(1, 4, 0x0408, 14)
20:32:41.989 NEO: modbus send(1, 4, 0x0500, 16)
20:32:42.241 NEO: modbus send(1, 4, 0x0600, 8)
20:32:42.493 NEO: modbus send(1, 4, 0x0022, 2)
20:32:42.749 NEO: modbus send(1, 4, 0x006A, 9)
20:32:42.995 HTP: Console
20:32:43.102 NEO: modbus send(1, 4, 0x0100, 17)
20:32:43.295 NEO: modbus send(1, 4, 0x0206, 20)
20:32:43.466 NEO: modbus send(1, 4, 0x0300, 7)
20:32:43.716 NEO: modbus send(1, 4, 0x0408, 14)
20:32:43.976 NEO: modbus send(1, 4, 0x0500, 16)
20:32:44.225 NEO: modbus send(1, 4, 0x0600, 8)
20:32:44.482 NEO: modbus send(1, 4, 0x0022, 2)
20:32:44.730 NEO: modbus send(1, 4, 0x006A, 9)
20:32:44.987 NEO: modbus send(1, 4, 0x0100, 17)
20:32:45.238 NEO: modbus send(1, 4, 0x0206, 20)
20:32:45.490 NEO: modbus send(1, 4, 0x0300, 7)
20:32:45.779 NEO: modbus send(1, 4, 0x0408, 14)
20:32:45.955 NEO: addr 0x0002 read data error 159
20:32:45.972 NEO: addr 0x0002 read data error 159
20:32:45.989 NEO: addr 0x0002 read data error 159
20:32:47.261 NEO: modbus send(1, 4, 0x0408, 14)
20:32:47.467 NEO: modbus send(1, 4, 0x0500, 16)
20:32:47.716 NEO: modbus send(1, 4, 0x0600, 8)
20:32:47.974 NEO: modbus send(1, 4, 0x0022, 2)
20:32:48.256 NEO: modbus send(1, 4, 0x006A, 9)
20:32:48.483 NEO: modbus send(1, 4, 0x0100, 17)
20:32:48.732 NEO: modbus send(1, 4, 0x0206, 20)
20:32:48.991 NEO: modbus send(1, 4, 0x0300, 7)
20:32:49.239 NEO: modbus send(1, 4, 0x0408, 14)
20:32:49.492 NEO: modbus send(1, 4, 0x0500, 16)
20:32:49.745 NEO: modbus send(1, 4, 0x0600, 8)
20:32:49.995 NEO: modbus send(1, 4, 0x0022, 2)
20:32:50.247 NEO: modbus send(1, 4, 0x006A, 9)
20:32:50.500 NEO: modbus send(1, 4, 0x0100, 17)
20:32:50.663 NEO: addr 0x0002 read data error 159
20:32:50.680 NEO: addr 0x0002 read data error 159
20:32:50.697 NEO: addr 0x0002 read data error 159
20:32:50.720 NEO: NeoPoolReadRegister(0x0429, 1) = [0x0000]
20:32:50.739 NEO: NeoPoolReadRegister(0x04EA, 1) = [0x0000]
20:32:51.977 NEO: modbus send(1, 4, 0x0100, 17)
20:32:52.227 NEO: modbus send(1, 4, 0x0206, 20)
20:32:52.487 NEO: modbus send(1, 4, 0x0300, 7)
20:32:52.734 WIF: Checking connection...
20:32:52.741 NEO: modbus send(1, 4, 0x0408, 14)
20:32:52.989 NEO: modbus send(1, 4, 0x0500, 16)
20:32:53.239 NEO: modbus send(1, 4, 0x0600, 8)

**Comment by curzon01** [comment]
As you can see your RS485 interface has communication issues, data expected from unit are not received.
Check the cables and powering for ESP and interfaces. Don't know how exactly you build your hardware. You need to remove the read errors. Are both LEDs on your RS485 interface are blinking frequently? Which of the Oxilife connector you are using and how the pins are connected? How do you power your RS485 and your ESP?

**Comment by ChristianBlenk** [comment]
Yes they do. All values in Tasmota are correct. Setpoints, voltages, relays
aso.

The Tasmota MCU and rhe RS485 module is soldered together. Perhaps the wire
to the Oxilife is too long. But RS485 is a serial bus for long distances.

Ill check this.

Norbert Richter ***@***.***> schrieb am Do., 29. Juni 2023,
21:59:

**Comment by curzon01** [comment]
> Yes they do. All values in Tasmota are correct. Setpoints, voltages, relays aso.

Bad interfacing may not lead to a complete non-functionality.

Values on main GUI indicates that communication is sometimes complete, but the read errors log is showing are not.
Polling of your Oxilife is done every 250 ms with a whole set of different Modbus registers each. A few of these may provide valid data, so you will see valid data, but your data refresh rate should also take longer than the normal 2 seconds due to communication errors  that occur.
Long cables on the RS485 side are usually no problem either. I use myself only a ~4 m simple Unitronic LiYY control cable 3 x 0.14mm² in a cable duct with other 230/400 V cables without any problems. Power supplying of the whole ESP/RS485 stuff is often given too little attention.

**Reply by ChristianBlenk** [reply]
Powersupply is 12V from Oxilife -> stepdown 5V -> EPS32 -> 3.3V -> RS485. Ill check all wires again. Probably a termination resistor is missing in the RS485 bus topology.
If i cant handle the problems i give esp8266 a try instead of esp32.

**Comment by curzon01** [comment]
termination resistor is not neccessray, powering the stuff using the Oxilife and stepdown could be the issue, I'd no luck in the past got it running stable, ESP32 is senstive. Using 8266 you will loose the berry support...

**Reply by ChristianBlenk** [reply]
How do you powering the stuff? 5V powersupply to esp and then 3,3v to rs485 or both vi 5v supply.

**Comment by curzon01** [comment]
ESP32 on 5V powersupply (MeanWell HDR-15-5), then 3,3V from ESP32 to RS485 (as shown in NeoPool doku [Connection](https://tasmota.github.io/docs/NeoPool/#connection))

Using 5V on RS485 may destroy your GPIOs earlier or later...

**Comment by ChristianBlenk** [comment, marked as answer]
Okay, done. TTL to UART converter was the problem. I think that the working voltage which was in the specs 3,3 to 5v is a mistake. It can only read/write single registers without timeout at low voltage. New ttl to uart do the job with stepdown voltage from the Oxylife. Thanks for support.
